ENH: EngFormatter new kwarg 'sep'

afvincent opened the pull request:
_Implements the proposition from #6533, by adding a kwarg `space_sep` to `ticker.EngFormatter`._

The default value of this new kwarg `space_sep` is `True` which preserves the current formatting, with a space separator between the value and the prefix/unit (like in "1.23 µV"). If  it is set to `False`, this space separator is removed (like in "1.23µV"). The latter is not really correct from the typographic viewpoint but will (for example) allow users who don't really care to save space in the ticklabels. As discussed in #6533, when no unit symbol is provided, "1.23 µ" and "1.23µ" seem to be both wrong (at least for typographers), so I chose to keep the current behavior as default.

The commit 0f37c41 fixes a minor inconsistency between different representations of zero. Now, `formatter(-0.0) == formatter(0) == formatter(-0) == u"0"` (with `formatter = EngFormatter()`), while before `formatter(-0) = u"0"` but `formatter(-0.0) = u"-0"` for example...

The commit 25cdf5f is a rewrite of the unit test `test_EngFormatter_formatting()` I previously added in PR #6014 . The new test now tests all the possible combinations of the kwargs, which was not the case before (in particular, only the default `places=None` was tested). Furthermore, the tested values now exercise a bigger range of cases (like for example 0, -0 and -0.0).

And by the way, I tried to improve a few docstrings. Hopefully I didn't add to many English mistakes or typos...

afvincent commented:
*Note to myself: don't forget to update the [example](http://matplotlib.org/examples/api/engineering_formatter.html) once some review will have been performed.*

tacaswell commented on lib/matplotlib/ticker.py:
```diff
@@ -1104,18 +1129,20 @@ def format_eng(self, num):
 
         formatted = format_str % (mant, prefix)
 
         formatted = formatted.strip()
```
I would remove this `strip`

tacaswell replied:
And move it up to L1088 (in the new code)

afvincent replied:
I think the solution in commit 2193c79 is more or less what you were proposing. 

In fact, the `strip` at the end of `format_eng` was the last operation done before the fix #6014, which added the if statement that appends a trailing space when there was no prefix and unit (to fix issue #6009). It seems to me that with the new `space_sep` option, this `if` block is not necessary anymore and that one simply needs to perform a `strip` in `__call__` before returning the final formatted string: the only case where it will remove a trailing space should the one without a unit and without a prefix. *The relevant test seems still OK with the new scheme.*

By the way, since #6014 (and only if `space_sep=True` since #6542), `format_eng(0)` returns `u"0 "` and not `u"0"` as it is written in the docstring. So, a) is this change an issue, since `format_eng` is not a private method?, b) should I fix the docstring?

tacaswell commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -452,23 +452,111 @@ def test_percentformatter():
+    DIGITS = set([u'{0:d}'.format(d) for d in range(10)])
+
+    # Case A/ With kwarg `places=None` (default)
+    A_wo_unit_w_sep = ((-1234.56789, u'-1.23457 k'),
```
:+1: to the very through testing

afvincent replied:
To be honest, the expected string in the cases `places=0` and `places=2` correspond to what was returned before this PR, but I find that some of the returned strings are weird. Like for example with `places=0`, 999.9 -> u'999' and 987654.321 -> u'987 k': shouldn't it be u'1000' , resp. u'988' instead? 

Besides, I'm not sure the trick with `DIGIT` is the cleanest one (or at least the more readable one).

QuLogic commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -452,23 +452,111 @@ def test_percentformatter():
+        """
+        eng_formatter = mticker.EngFormatter(places=places, space_sep=False)
+        # Remove the space separator from the reference cases
+        test_cases = ((val, u"{s:s}".format(s=s.replace(" ", "")))
```
Is the string format really needed here?

afvincent replied:
I simply copy-pasted from one of the previous similar case and didn't pay more attention, but you are right: it's clearer without the string format. Removed by commit  8d086e0.

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1058,9 +1061,27 @@ class EngFormatter(Formatter):
 
-    def __init__(self, unit="", places=None):
+    def __init__(self, unit="", places=None, space_sep=True):
+        """ Parameters
```
I think this should be on the next line; I'm not sure the leading space trimming will work properly like this.

afvincent replied:
Commit 4d96e47 should have modified all occurences of
```python 
""" Text
"""
```
into
```python
"""
Text
"""
```

afvincent commented:
Not sure the Travis failure is related to my last commit.

afvincent commented:
BTW, what should be done about the problem noticed l. 1134 in `ticker.py` (`# TODO: shouldn't we raise a warning if self.places < 0?`)? Should an explicite exception be raised, or do we assume users won't call an instance of EngFormatter with a negative value for the kwarg `places`.

afvincent commented:
I rewrote the example about the EngFormatter in the API section (`engineering_formatter.py`) to demonstrate about the new option `space_sep`. Eventhough most of the lines have been modified, it is the same spirit and data as before (the former example is kept as left panel). 

I simply moved the demo of the kwarg `places` from the former example (left panel) to the new one (right panel) for esthetical reasons: with `places=1` in the left panel, the tick labels were a bit mixed.

Here is a example how it now looks like:
![updated_example](https://cloud.githubusercontent.com/assets/17270724/16176932/c9b70e1c-361d-11e6-9425-0f032d09798f.png)

QuLogic commented:
Yes, the nose stuff needs to be removed now.

afvincent commented:
I will try to update to pytest tomorrow (“later today” may be more correct).

afvincent commented:
I rebased my local copy on the current master (things did not go totally smooth). I removed the changes that I had made in `test_ticker` because they were conflicting with the ones made during the Pytest migration. And of course, I have issues with Pytest when I want to check if I broke anything (it is complaining about `cycler`, which is installed and up-to-date on my system though).

So I will wait for CI to finish running on the rebased version before re-adding my previous changes in `test_ticker`, converted to Pytest.

afvincent commented:
Huh, what did I do wrong with `rebase` to get twice my GitHub thumbnail with commits 43d812d, 9121205, a1a38c5 and 7bad242 ^^?

dopplershift commented:
I wouldn't call it wrong, but I bet you have a new email set on github; the second icon is for the person who committed the change, rather than the author.

afvincent commented:
As Travis seems to be fine with the rebase (and the AppVeyor failure does not look like related), I have just (re-)included to extended test of EngFormatter that I had introduced before (If I did not forget any, it should of course include all the test cases that were tested in the previous commit). I hope I translated it well in Pytest style: we will see if CI (at least Travis) is still happy with this new version!

@dopplershift I think you may be write for the double thumbnail (I remind pushing stuff from computers with different email addresses). Well I guess that in the end, we will play with the mailmap file if it does mess with the statistics.

afvincent commented:
Well, at least both CI services agree on failing \o/… From a rapid glance at the logs, it seems that I mixed some “expected outputs” in the test revamping. I will double check and take care of it.

Tillsten commented:
Why not use a half space or make the space character customizable?

afvincent commented:
Well I think it just seemed easier to me to implement this with a boolean flag when I opened this PR (I was more or less a newcomer to mpl/git/test suite & Co).

What you suggest may indeed be better from the user viewpoint. We could replace the kwarg `space_sep=True(default)|False` with `separator=u" " (default)|any string`. *Because backward compatibility is something we worship, the new default would still have to be a full space* (not even unbreakable :/) as it is the original behavior. However, you could remove the space separator with `separator=""` or even use a narrow no-break white space with `separator=u'\u202f'` if you prefer.

@tacaswell @QuLogic @phobson @dopplershift (summonning every people who posted in the thread ^^) Thoughts? Who this idea be better than the current one?

afvincent commented:
BTW, just `sep` may be a better name for a customizable separator argument. For example, I just discovered that `print` also has a separator argument, called `sep`. (Always learning…)

tacaswell commented:
:+1: to `sep` as the kwarg name and :+1: to it being configurable rather than just a boolean.

:-1: to a non-ascii default, but :+1: to mentioning that in the docs.

Tillsten commented:
What wrong with an non ASCII default?

afvincent commented:
@tacaswell Just to be sure, by non-ascii default, you mean that it should be `sep=" "` instead of `sep=u" "`?

tacaswell commented:
I mean `" "` instead of `'\u202f'`.

You should not need the leading `u` anywhere.  We have `unicode_literals` at the top of all of our files.  We no longer support 3.2 so they are not _forbidden_ anymore, but just not needed.

afvincent commented:
Ok thank you for explaining! I will try to implement this idea tomorrow. (And of course my plan was to use a full white space as default :P)

afvincent commented:
Argh! I forgot to update the docstring of the example… I will correct this ASAP.

afvincent commented:
I think everything is now quite OK for (re-)review, with the new behavior based on @Tillsten 's suggestion. Maybe the test could have been more factorized (but I wanted to keep the testing of (almost) default case apart).

tacaswell commented on lib/matplotlib/ticker.py:
```diff
@@ -1212,12 +1216,32 @@ class EngFormatter(Formatter):
+            String used between the value and the prefix/unit. Beside the
+            default behavior, some other useful use cases may be:
+            * sep="" to append directly the prefix/unit to the value;
+            * sep="\\u00a0" to use a no-break space;
```
Why is a no-break space useful here?  You have to go way out of your way to get mpl to try to wrap your text.

afvincent replied:
I can remove this line if you think it is not that useful. I added the information just because (at least) in French typography, one are supposed to use a no-break space before units. And sometimes, I directly use an EngFormatter instance to format a value and its unit in a longer string (that is not a tick label). So I guess there might be some edges cases where unfortunate wrapping could happen.

afvincent replied:
Well, on second thought, I will at least switch lines between the regular no-break- and the narrow no-break spaces, in order to show the cases in decreasing order of “usefulness” for most people.

tacaswell replied:
Interesting what one can do when free from ascii.

Had not considered that this would be used outside of a plot context.

tacaswell commented:
:+1: This probably deserves an entry in whats_new?

tacaswell commented on examples/api/engineering_formatter.py:
```diff
@@ -14,13 +14,29 @@
+# Demo of the options `places` (number of digit after decimal point) and
+# `sep` (separator between the number and the prefix/unit).
+ax1.set_title('SI-prefix only ticklabels, 1-digit precision & no separator')
+formatter1 = EngFormatter(places=1, sep="")
```
maybe show the narrow-space here?

afvincent replied:
I will adapt the example (I was hesitating doing it during the docathon ^^)

QuLogic commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -524,26 +524,93 @@ def test_basic(self, format, input, expected):
+
+        # Case 0: unit='' (default) and sep=' ' (default).
+        # 'expected' already corresponds to this reference case.
+        exp_outputs = (_s for _s in expected)  # simple copy of 'expected'
```
This is not really a copy; it's a generator expression.

afvincent replied:
You're right.  I think I will simply remove the comment. In fact, I am wondering if this generator is useful in this very case. `expected` is a tuple, so it is not mutable: simply using `exp_outputs = expected` should be safe and not risk to alter `expected`.

QuLogic commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -524,26 +524,93 @@ def test_basic(self, format, input, expected):
+
+            # Case 3x: unit='' (default) and sep=_sep.
+            # Remove the space separator from the reference case and append
+            # a unit symbol to it.
```
I thought this case had no unit?

afvincent replied:
Yes, you are right. The cases 2. and 3. were switched during a previous commit, and I failed at (completely) switching the relevant comments… I will take care of it.

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1212,12 +1216,32 @@ class EngFormatter(Formatter):
+        """
+        Parameters
+        ----------
+        unit: string (default: "")
```
Needs a space before the colon. Also, the type is `str`.

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1212,12 +1216,32 @@ class EngFormatter(Formatter):
+            If it is None, falls back to the floating point format '%g'.
+
+        sep: string (default: " ")
+            String used between the value and the prefix/unit. Beside the
```
Beside -> Besides

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1212,12 +1216,32 @@ class EngFormatter(Formatter):
+
+        sep: string (default: " ")
+            String used between the value and the prefix/unit. Beside the
+            default behavior, some other useful use cases may be:
```
These are not use cases; they are useful values/options/etc..

QuLogic commented on lib/matplotlib/ticker.py:
```diff
@@ -1212,12 +1216,32 @@ class EngFormatter(Formatter):
     def __call__(self, x, pos=None):
         s = "%s%s" % (self.format_eng(x), self.unit)
+        # Remove the trailing separator when there is neither prefix nor unit
+        s = s.strip(self.sep)
```
This does not work the way you want. `strip` removes _any_ character in its argument from both sides: `'abcaba'.strip('aba') = 'c'`. Say someone uses `sep='0'` (for some reason), then `value = '0.00'` + ` sep = '0'` would be `'0.000'`, and it would become `'.'`.

afvincent replied:
Indeed, I did not think of this case. I will replace it with a simple
```python
if s.endswith(self.sep):
    s = s[:-len(self.sep)]
```

tacaswell replied:
That also needs a cut-out for zero length `sep` as `'' == 'abce'[:-0]`

afvincent commented on examples/api/engineering_formatter.py:
```diff
@@ -14,13 +14,29 @@
 xs = np.logspace(1, 9, 100)
-ys = (0.8 + 0.4 * prng.uniform(size=100)) * np.log10(xs)**2
-ax.plot(xs, ys)
+ys = (0.8 + 0.4*prng.uniform(size=100))*np.log10(xs)**2
```
Huh, I guess it is an aftermath of my windy rebase, during which I had to take care of conflicts with another PR. I will go back to the previous spacing.

afvincent commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -524,26 +524,93 @@ def test_basic(self, format, input, expected):
+        argument, in several cases:
+            0.  without a unit symbol but with a (default) space separator;
+            1.  with both a unit symbol and a (default) space separator;
+            2x. with both a unit symbol and some non default separators;
```
I was using it to indicate that both cases 2. and 3. were looped over several subcases ("2a", "2b", etc. in my head…). But the meaning is not obvious, so I will just switch for "2." and "3."

afvincent commented:
I will take care of @QuLogic 's remarks later in the day. I need to sleep a bit or there will be more things to fix *after* the relevant commits ;).

afvincent commented:
All the previous comments should have been addressed.

My remaining concern, which are rather minor, are that in the HTML documentation based on the docstrings of the `EngFormatter` class:

1. the dictionary of the SI prefixes (`ENG_PREFIXES`) is interpreted as something to appear, eventhough IMO it is not supposed to. But it is not dramatic and I guess it may be OK to go with the current state.
2. the unicode character strings are not displayed correctly. For example, '\\u202f' becomes 'u202f'… This 2nd point may be more problematic than the 1st one :/.

The relevant HTML documentation currently looks like
![engformatter_docstring_new_version](https://cloud.githubusercontent.com/assets/17270724/22875927/80e367cc-f1ce-11e6-9e5a-fbed71c7b0e1.png)

QuLogic commented:
Put the `sep` parts in double-backticks so they appear as code and Sphinx doesn't try to interpret them.

afvincent commented:
@QuLogic Thanks, it seems to be OK now :). The current output is
![engformatter_docstring_with_double-backticks](https://cloud.githubusercontent.com/assets/17270724/22892678/b1c99452-f213-11e6-9553-ebabb9b3e62a.png)

NelleV commented:
@afvincent can you rebase? The rest looks good.

anntzer commented on doc/users/whats_new/EngFormatter_new_kwarg_sep.rst:
```diff
@@ -0,0 +1,10 @@
+New keyword argument 'sep' for EngFormatter
+-------------------------------------------
+
+A new “sep” keyword argument has been added to
```
Use regular, not curly quotes here?

afvincent replied:
Fixed by 5feb554

anntzer commented on examples/api/engineering_formatter.py:
```diff
@@ -14,13 +14,30 @@
+# `sep` (separator between the number and the prefix/unit).
+ax1.set_title('SI-prefix only ticklabels, 1-digit precision & ' +
+              'hair space separator')
+formatter1 = EngFormatter(places=1, sep=u"\u200a")  # U+200A is hair space
```
You should use a thin space (https://en.wikipedia.org/wiki/Thin_space).  "\N{THIN SPACE}" (or "\N{HAIR SPACE}") is also more self-documenting.

afvincent replied:
Fixed by 8616a04

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1212,12 +1205,40 @@ class EngFormatter(Formatter):
+            other useful options may be:
+
+            * ``sep=""`` to append directly the prefix/unit to the value;
+            * ``sep="\\u2009"`` to use a thin space;
```
Again, named entities ("\N{...}") are more self-documenting.

afvincent replied:
Fixed by 8616a04

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1258,19 +1283,17 @@ def format_eng(self, num):
 
         mant = sign * dnum / (10 ** pow10)
 
+        # TODO: shouldn't we raise a warning if self.places < 0?
```
I think not mixing two formatting methods would be more legible, e.g.
```
formatted = "{mant:{fmt}}{sep}{prefix}".format(
    mant=mant, prefix=prefix, sep=sep,
    fmt="g" if self.places is None else "d" if self.places == 0 else ".{}f".format(self.places))
```
(possibly moving the last line out, but note that the ability to nest formatting expressions is a real plus for str.format here...)

afvincent replied:
More or less done, but there is still the issue of 2 remaining failing tests (see my comment below).

QuLogic commented:
Is there much missing here? Maybe just a rebase to see what the docs look like?

afvincent commented:
@QuLogic I will try to address @anntzer remark tonight or tomorrow morning. I will rebase when I will be there ;).

afvincent commented:
@anntzer Stay tuned, I normally just fixed your comments. Still have to build the doc locally to check that everything is still OK, and if so, I will force-push for review ;).

afvincent commented:
**Do not merge 13f333d !**

Beware that it should fail on 2 (and only 2...) peculiar tests, both with `places=0`:
  * when value is 999.9, the new version returns "1000" while the test expects "999";
  * when value is 987654.321, the new version returns "988 k" while the test expects "987 k".

I am pretty sure that when I wrote the tests in the first places I expressed doubts about the expected values, which were based on the original behavior. Unfortunately, I cannot find the comment or post where I think I did this :(. I am pretty convinced that these former values are not the ones that would genuinely be expected, like "1 k" and "988 k". But maybe I am just biased.

afvincent commented:
@anntzer So you think (from what you wrote on Gitter) that I had better stop struggling with `decimal.Decimal` and just try to re-write the stuff with genuine scalars and standard types? Disclaimer: the class was already using `decimal` when I tweaked it, so I do not really know the ground reason for really on it in `EngFormatter`.

tacaswell commented:
:+1: on ripping out decimal and both of those roundings seem wrong.

afvincent commented:
@tacaswell That is what I thought in the first place, but on 2nd thought, it would be better if 999.9 was actually resulting in "1 k" rather than "1000", doesn't it :/ ?

afvincent commented:
Well anyway, if we are going to get rid of `decimal`, I will have some significant changes to do => I am going to remove the [MRG+1].

QuLogic commented:
I'm not sure we _need_ to rewrite everything just to get this in; it seems like it's doable without all that work.

anntzer commented:
I am totally fine with the "remove decimal" part going in a separate PR.

afvincent commented:
Well it looks like it was easier than I expected to switch from `decimal.Decimal` for simple genuine floats.

All the tests were passing locally (Python 2.7) except the two "unnatural" ones. So I went further and fixed theses 2 tests. Please note that the formatter now handles correctly (AFAIAC) the corner-case of 999.9... being rounded to 1000 which then requires to update the SI prefix.
@tacaswell Should it be stated somewhere that these two last points have changed?

*I think it is now ready for final review.* I would not want to appear selfish or pressing here but I would be very glad if this PR could make it into the incoming release (if everything is good enough of course). I am regularly missing this functionality, and it seems like I am actually not alone (#9042) ;).

afvincent commented:
Let's see if I can get rid of the exception `ValueError: Integers to negative integer powers are not allowed.` Weirdly, it did not occur during my local tests.

tacaswell commented:
That is a newer-version of numpy thing.  The issue is that in general and int raised to a negative power is going to result in a float.  Numpy could just do that and return a float array, but then `a2 = a ** (-2)` and `a **= (-2)` give different answers as in the second case the values would be cast back to int and rounded to 0.

afvincent commented:
@tacaswell Indeed, I just found [this](https://docs.scipy.org/doc/numpy/release.html#power-and-raise-errors-for-integer-to-negative-integer-powers).

afvincent commented:
Apart from one of the coverage tests (and I do not understand why this peculiar one is failing), the other tests are passing :). So pinging @anntzer and @tacaswell for review if they want to/can have a look.

anntzer commented on lib/matplotlib/tests/test_ticker.py:
```diff
@@ -549,26 +549,95 @@ def test_basic(self, format, input, expected):
+
+        # Test several non default separators: no separator, a narrow
+        # no-break space (unicode character) and an extravagant string.
+        for _sep in ("", "\u202f", "@_@"):
```
`\N{THIN SPACE}` here too?  (not a big deal)

afvincent replied:
Fixed in ec92520

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1238,40 +1258,39 @@ def format_eng(self, num):
 
         `num` may be a numeric value or a string that can be converted
-        to a numeric value with the `decimal.Decimal` constructor.
+        to a numeric value with `float(num)`.
```
should use double backquotes (single backquotes are for things that refer to other entries in the docs).

afvincent replied:
Fixed in ec92520

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1214,12 +1206,40 @@ class EngFormatter(Formatter):
+            Precision with which to display the number, specified in
+            digits after the decimal point (there will be between one
+            and three digits before the decimal point). If it is None,
+            falls back to the floating point format '%g'.
```
"%g" defaults to a precision of 6.  Does that mean that the default is simply places=6?  Otherwise the statement is a bit unclear.

afvincent replied:
Actually, if I am correct, "%g" results in up to 6 *significant* digits. This means that with it, the equivalent value for ``places`` will be between 0 and 5. I'll try to make the docstring clearer.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1238,40 +1258,39 @@ def format_eng(self, num):
+        mant = sign * dnum / (10.0 ** pow10)
+        # Taking care of the cases like 999.9..., which
+        # may be rounded to 1000 instead of 1 k.
+        if (self.places is not None and
```
See above as to places == None: is it clear that we do not want to fix the rounding in that case too?

afvincent replied:
I have another idea on how to handle this specific case, which would also be abe to take care of the "%g" format case. A dedicated commit is on its way.

afvincent replied:
With commit  e55990d, the idea is now simply to format the string and to convert it back to a float to see if the value ends up in the range of the next SI prefix. This way, even "%g" and its non-fixed value for `places` should be correctly handled.

BTW, being there, I also fixed the case where the value to format is way to big (>= "1000 Y").

afvincent commented:
@anntzer I hope everything is fine this time. Let's first see if the CI services agree :).

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1214,12 +1206,43 @@ class EngFormatter(Formatter):
+            and three digits before the decimal point). If it is None,
+            the formatting falls back to the floating point format '%g',
+            which displays up to 6 *significant* digits (i.e. the
+            equivalent value for *places* varies between 0 and 5, both
```
between 0 and 5 (inclusive)
(perhaps a native speaker can comment on this)

anntzer replied:
Also, now that I think of it, this behavior (which was already present before, I know) is a bit weird, isn't it?  `f` also defaults to 6 decimal places (not significant digits, I know) (see `pydoc FORMATTING`), so basically the possible formats are

.6g (== g) if places == None
.0f .1f .2f .3f etc. if places == 0 1 2 3 etc.

It would probably be a bit awkward to change this, but it looks like in an idea API places should be set by one parameter (perhaps defaulting to 6) and f vs g by another.

afvincent replied:
Fixed in 3bedd6e.

afvincent replied:
If we want to go this way, I´d rather just replace the `places` argument with something like `format`, which would defaults to `.6g` (or `g` if I understand correctly). But as you warned, it would definitively be a break of the current API ;).

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1238,40 +1261,41 @@ def format_eng(self, num):
 
         `num` may be a numeric value or a string that can be converted
-        to a numeric value with the `decimal.Decimal` constructor.
+        to a numeric value with ``float(num)``.
```
Do we actually have a use case for supporting this?  Could this behavior be deprecated?  (this is not strictly needed for the PR)

afvincent replied:
String were accepted before so I let it here. But other formatter classes do not seems to support such a feature, so if one thinks it is not worth supporting it anymore, it could be deprecated indeed.

afvincent replied:
**Note to myself:** add a deprecation warning for 2.1 (=> removing in 2.3?).

afvincent replied:
Deprecated in f207c89

anntzer commented on examples/api/engineering_formatter.py:
```diff
@@ -14,13 +14,31 @@
+
+# Demo of the options `places` (number of digit after decimal point) and
+# `sep` (separator between the number and the prefix/unit).
+ax1.set_title('SI-prefix only ticklabels, 1-digit precision & ' +
```
no need for the +

afvincent replied:
Fixed in 3bedd6e.

afvincent commented:
I added the deprecation that @anntzer suggested (and added an entry in the `api_changes` folder).

I rebased and being there I squashed some small commits (still 25 commits though...). I struggled a bit with a conflict due to the recent deletion of  `deprecations.rst` in `api_changes` on master, but I managed to find a way around (I hope it will be fine).

TBH, on @anntzer's last comment about the parameter `places` and the “weird” formatting behavior, I would prefer this to be the matter of another (i.e. future) PR, *if one eventually decides to change the way of interacting with the formatter formatting*.

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -1214,12 +1206,42 @@ class EngFormatter(Formatter):
 
     def __call__(self, x, pos=None):
         s = "%s%s" % (self.format_eng(x), self.unit)
+        # Remove the trailing separator when there is neither prefix nor unit
```
This could be rewritten as `s = re.sub(re.escape(self.sep) + "$", "", s)` (without the need for a conditional) but I'm not going to block the PR on the fix.

afvincent replied:
Well regex are black magic to me, but as you kindly gave the answer plus the tests still passed locally, here it is with 033111f ;).

WeatherGod commented:
@HDembinski has noted that the use of regex in the last commit significantly slowed down this bit of code. Let's roll back that change as the use of the regex doesn't really justify the performance penalty.

afvincent commented:
@WeatherGod I just rebased and dropped the last commit, which introduced the use of a regex.

@HDembinski Thank you for benchmarking :).

afvincent commented:
I do not understand what is going wrong with AppVeyor :/. I already restarted it twice and it always fails at during the same build (Python 3.5) with what look like network issues
```
CondaError: CondaHTTPError: HTTP 000 CONNECTION FAILED for url <https://repo.continuum.io/pkgs/free/win-64/mkl-2017.0.3-0.tar.bz2>
Elapsed: -
An HTTP error occurred when trying to retrieve this URL.
HTTP errors are often intermittent, and a simple retry will get you on your way.
```
But three times in a row, it is not really intermittent to me anymore ^^...

afvincent commented:
Marking it as release-critical because it is a feature that I would really appreciate to see in next release and I think it is finally in good shape (apart from the AppVeyor failure maybe). Feel free to remove the label if you disagree.

tacaswell commented:
:tada:

afvincent commented:
@tacaswell Thank you :).

And many thanks to all the people who got involved and reviewed this PR!
